tuned/plugins/plugin_net.py

```diff
@@ -337,15 +369,17 @@ def _check_device_support(self, context, parameters, device, dev_params):
 			parameters.pop(param, None)
 
 	def _get_device_parameters(self, context, device):
-		context2opt = { "coalesce": "-c", "features": "-k", "pause": "-a", "ring": "-g" }
+		context2opt = { "coalesce": "-c", "features": "-k", "pause": "-a", "ring": "-g", \
+                        "channels": "-l"}
 		opt = context2opt[context]
 		ret, value = self._cmd.execute(["ethtool", opt, device])
 		if ret != 0 or len(value) == 0:
 			return None
 		context2parser = { "coalesce": self._parse_device_parameters, \
 				"features": self._parse_device_parameters, \
 				"pause": self._parse_pause_parameters, \
-				"ring": self._parse_ring_parameters }
+				"ring": self._parse_ring_parameters, \
+				"channels": self._parse_channels_parameters }
```

Comment:
A few questions so I can better understand the implications:
What happens if a device needs less queues then specified in the tuned profile?
What happens if the number of queues in the tuned profile is less than minimum queues needed by the device to work properly?
What happens if the number of IRQ vectors per CPU gets exhausted because of too many SR-IOV functions or other devices?
Is this the right place to check/ensure the above?

Reply:
Maybe a per device settings would also make sense? Or the setting is per device and I missed that?

Reply:
> A few questions so I can better understand the implications:
> What happens if a device needs less queues then specified in the tuned profile?

Tuned will still go ahead and allocate the queues as per the specification in the tuned profile.

The general trend from the device perspective is to have 1 queue per available CPU and that's what we are following here. Tuned should not automatically reduce anything to a minimum as that will impact the performance.

However, if a user wants more queues then tuned should not prevent that either.

> What happens if the number of queues in the tuned profile is less than minimum queues needed by the device to work properly?

This kind of configuration should automatically fail.
The same is the case if you specify a number that is greater than the max allowed queues for the device, tuned/ethtool will simply return an error without any change.

> What happens if the number of IRQ vectors per CPU gets exhausted because of too many SR-IOV functions or other devices?

Do you mean something like a block device? if so then yes that is a control we would like to integrate into tuned as well. However, that can not be achieved from ethtool, so first we have to find a tool that can do that, and then we will have to integrate its support in tuned.
As the first step, this patch-set is only taking care of the network devices.

> Is this the right place to check/ensure the above?

So, one change that I am planning to do in the next version or as another patch set is to terminate tuned execution if tuned's scheduler plugin fails due to ENOSPC (suggested by Marcelo). That way the user will atleast know that some device (eg block) queues need to be adjusted manually since tuned can only take care of the network device queues at the moment.

I also hope that in the future we can have something more concrete from the kernel side to prevent these kinds of situations.

Please let me know if you have any questions/comments.

Reply:
> Maybe a per device settings would also make sense? Or the setting is per device and I missed that?

Per device setting that can be controlled by the user? 
The idea is to let tuned configure the device (network for now) queues generically based on HK CPUs for RT setup similar to what kernel does base on the online CPUs at the time of device initialization. 

If the queues for each device need to be adjusted individually then the user can still do that through ethtool directly.

Reply:
The issue here is that not some workloads use kernel networking (as opposed to DPDK), so we want to make sure that it is still possible to configure the queues based on workload knowledge to use enough cpus for interrupt handling.

Reply:
And the per device idea is something similar. There are cases where some NICs are used for DPDK (no queues or interrupts needed really) and some others are used for less important work (management, reporting, aggregated data..). It might be ok to reduce the mgmt NICs queues to housekeeping cores too as the traffic is usually light. But it is an use case that needs to be taken into account.

Reply:
> The issue here is that not some workloads use kernel networking (as opposed to DPDK), so we want to make sure that it is still possible to configure the queues based on workload knowledge to use enough cpus for interrupt handling.

May I know who is responsible for controlling/adjusting the queues for these workloads right now based on the workload requirement? I would assume this is still done via userspace using something like ethtool?

Reply:
Yep, ethtool in userspace.

What we are after here is some kind of tuned variable that would let the user override the defaults if needed. I know it is tricky to do this per NIC though..

Reply:
> Yep, ethtool in userspace.
> 
> What we are after here is some kind of tuned variable that would let the user override the defaults if needed. I know it is tricky to do this per NIC though..

It should be doable but will require certain changes in plugin_net.py.
Right now, all ethtool commands are integrated in a way that the changes are applied to all the networking devices.

A very crude way that I can think of for doing this is by introducing another variable such as netdev_name.
With this we will have a few options to configure for eg in realtime-virtual-host.conf:
netdev_name=all or netdev_name=eth12 or netdev_name=eth12,eth13,...

realtime's tuned.conf will change to:
[net]
channels=${netdev_name} combined ${f:get_queue_count:${netdev_queue_count}}

The usability of netdev_queue_count will still remain as it is proposed in this patch-set.
What do you think?

Reply:
That might work indeed

Reply:
Thanks for confirming.
@yarda any initial thoughts or objections on making this change?

I can start working on this and send another version of the patch.
The only detail here is that I may have to either set "netdev_name=all" as the default setting in all the configurations that inherit the realtime profile or else I can just add a comment suggesting to set this in the realtime-virtual-host.conf, and other files similar to what we have for isolated_cores.